### PR TITLE
Improve typed array support.

### PR DESCRIPTION
**Pull request by Ms2ger**
This change is [Reviewable](https://reviewable.io/reviews/servo/rust-mozjs/304)

### CODE REVIEW

**Comment by jdm** [comment]
I am super pumped to look at this later today!

File: src/typedarray.rs [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+/// Internal trait for creating new typed arrays.
+pub trait TypedArrayElementCreator: TypedArrayElement {
+    /// Create a new typed array.
+    unsafe fn create_new(cx: *mut JSContext, length: u32) -> *mut JSObject;
+    /// Get the data.
+    unsafe fn get_data(obj: *mut JSObject) -> *mut Self::Element;
+}
+
+macro_rules! typed_array_element {
```
Comment (jdm) [inline src/typedarray.rs]: I propose one macro branch that accepts $t, $element, $unwrap, and $length_and_data, which declares the struct and implements TypedArrayElement, and another branch that accepts all the existing arguments, invokes the previous branch, and implements TypedArrayElementCreator.

File: src/typedarray.rs [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+                     UnwrapArrayBuffer,
+                     GetArrayBufferLengthAndData,
+                     JS_NewArrayBuffer,
+                     JS_GetArrayBufferData);
+
+
+/// The ArrayBufferView type.
+pub struct ArrayBufferViewU8;
+
+impl TypedArrayElement for ArrayBufferViewU8 {
```
Comment (jdm) [inline src/typedarray.rs]: We shouldn't need this if we modify the typed_array_element macro as suggested above.

**Comment by jdm** [comment]
As for the neutering stuff, I've spent a while looking at Gecko's implementation and thinking about whether we can do anything better, and I haven't come up with any good ideas. APIs like TypedArray::with_slice could accept a closure, but that wouldn't prevent people from running JS that could invalidate the slice inside of the closure, so it would still be unsafe and less pleasant to use.

**Comment by Ms2ger** [comment]
Still needs tests.

**Comment by Ms2ger** [comment]
Tests done, r? @jdm

**Comment by jdm** [comment]
@bors-servo: r+
Thanks!

**Pull request merged**